### PR TITLE
feat: update k8s deps, Go version and grafana operator version

### PR DESCRIPTION
**Pull request by vmanley**
Jira: [MGDSTRM-9968](https://issues.redhat.com/browse/MGDSTRM-9968)

This change will update a number of dependencies including:
* k8s.io/api => v0.24.3
* k8s.io/apimachinery => v0.24.3
* k8s.io/apiserver => v0.24.3
* k8s.io/client-go => v0.24.3
* sigs.k8s.io/controller-runtime => v0.12.2
* Go => 1.19
* Grafana Operator => v4

These dependency updates are necessary to ensure the Operator to support OpenShift 4.12.

This change will also deploy Prometheus v2.36.2 by default

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,57 +1,167 @@
...
-	github.com/opencontainers/image-spec => github.com/opencontainers/image-spec v1.0.2
-	github.com/opencontainers/runc => github.com/opencontainers/runc v1.1.2
-	go.mongodb.org/mongo-driver => go.mongodb.org/mongo-driver v1.5.1
-	k8s.io/api => k8s.io/api v0.19.2
-	k8s.io/apimachinery => k8s.io/apimachinery v0.19.2
-	k8s.io/apiserver => k8s.io/apiserver v0.19.2
-	k8s.io/client-go => k8s.io/client-go v0.19.2
-	sigs.k8s.io/controller-runtime => sigs.k8s.io/controller-runtime v0.6.3
-	sigs.k8s.io/controller-runtime/pkg/client/fake => sigs.k8s.io/controller-runtime/pkg/client/fake v0.12.1
+	k8s.io/api => k8s.io/api v0.24.3
```
Comment (MStokluska) [inline go.mod:160]: I think most of these replaces can be removed

File: Dockerfile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Build the manager binary
-FROM registry.access.redhat.com/ubi8/go-toolset:1.17.10 as builder
+FROM golang:1.18 as builder
```
Comment (MStokluska) [inline Dockerfile:2]: would it be better to use `registry.ci.openshift.org/openshift/release:golang-1.18 AS builder`

File: go.mod (hunk 1 above)
Comment (vmanley) [reply go.mod:160]: Thanks @MStokluska I'll revise this replace list

File: Dockerfile (hunk 2 above)
Comment (vmanley) [reply Dockerfile:2]: Not sure of benefits here. Do you have a link to this image?

File: go.mod [hunk 3]
```diff
@@ -1,57 +1,167 @@
...
-	github.com/prometheus-operator/prometheus-operator v0.43.0
-	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.43.0
+	github.com/prometheus-operator/prometheus-operator v0.50.0
+	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.50.0
 	github.com/sirupsen/logrus v1.8.1
-	k8s.io/api v0.20.6
-	k8s.io/apimachinery v0.20.6
+	k8s.io/api v0.24.3
+	k8s.io/apimachinery v0.24.3
 	k8s.io/client-go v12.0.0+incompatible
```
Comment (MStokluska) [inline go.mod:21]: hey @vmanley; I think the client-go will also have to be bumped to 0.24.3

Comment (vmanley) [reply go.mod:21]: @MStokluska Hey sorry missed this comment. That's no problem

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -16,7 +16,7 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@v2
       with:
-        go-version: 1.16
+        go-version: 1.18
```
Comment (machi1990) [inline .github/workflows/go.yml:19]: 1.19 according to https://github.com/redhat-developer/observability-operator/blob/16c9ed525e38d7f4ad2e7d5ee8f80f0ec2f4bda7/go.mod#L3?

File: .github/workflows/pr-check.yaml [hunk 5]
```diff
@@ -13,7 +13,7 @@ jobs:
       - name: Setup go
         uses: actions/setup-go@v2
         with:
-          go-version: "1.16"
+          go-version: "1.18"
```
Comment (machi1990) [inline .github/workflows/pr-check.yaml:16]: same comment as https://github.com/redhat-developer/observability-operator/pull/151/files#r1032274171

File: .github/workflows/go.yml (hunk 4 above)
Comment (vmanley) [reply .github/workflows/go.yml:19]: Thanks @machi1990 I'll update that

File: go.mod [hunk 6]
```diff
@@ -1,64 +1,93 @@
...
-	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.55.1
+	github.com/prometheus-operator/prometheus-operator v0.58.0
+	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.58.0
 	github.com/prometheus/client_golang v1.13.0
 	github.com/sirupsen/logrus v1.8.1
-	k8s.io/api v0.23.5
-	k8s.io/apimachinery v0.23.5
+	k8s.io/api v0.24.3
+	k8s.io/apimachinery v0.24.3
 	k8s.io/client-go v12.0.0+incompatible
```
Comment (MStokluska) [inline go.mod:23]: Hey, looks good to me but I would remove the replaces `replace k8s.io/client-go => k8s.io/client-go v0.24.3` and update it directly in require

Comment (vmanley) [reply go.mod:23]: Thanks @MStokluska. I had to revert back to using replace here after rebasing after another PR. I'll check again if we can avoid the replace

Comment (vmanley) [reply go.mod:23]: @MStokluska Don't think we can get rid of the replace here. We need to support install of [cluster-logging-operator](https://github.com/openshift/cluster-logging-operator) which is using `k8s.io/client-go v12.0.0+incompatible`

File: controllers/reconcilers/configuration/configuration_reconciler.go [hunk 7]
```diff
@@ -228,6 +230,18 @@ func (r *Reconciler) Reconcile(ctx context.Context, cr *v1.Observability, s *v1.
...
+	// Force resync if any CRs missing post-migration
+	if s.Migrated {
+		var missingCR bool
+		missingCR, err = r.checkForMissingCR(ctx, cr)
+		if err != nil {
+			return v1.ResultFailed, errors2.Wrap(err, "error checking for migrated CR")
+		}
+		if missingCR {
+			overrideLastSync = true
+		}
```
Comment (MikeEdgar) [inline controllers/reconcilers/configuration/configuration_reconciler.go:242]: @vmanley , @pb82 - does this need to consider the value of `.spec.descopedMode.enabled`? I'm seeing very frequent resyncs with descopedMode and only a Prometheus CR is present.

Comment (vmanley) [reply controllers/reconcilers/configuration/configuration_reconciler.go:242]: @MikeEdgar Sounds like that would cause an issue. I'll look into it

**Pull request merged**